src/metabase/db/custom_migrations.clj

```diff
@@ -352,3 +361,88 @@
+    (run! update-one! (t2/reducible-query {:select [:id :object]
+                                           :from   [:revision]
+                                           :where  [:and
+                                                    [:= :model "Card"]
+                                                    [:or
+                                                     ;; these match legacy field refs in column_settings
+                                                     [:like :object "%ref\\\\\",[\\\\\"field-id%"]
+                                                     [:like :object "%ref\\\\\",[\\\\\"field-literal%"]
+                                                     [:like :object "%ref\\\\\",[\\\\\"fk->%"]
+                                                     ;; MySQL with NO_BACKSLASH_ESCAPES disabled:
+                                                     [:like :object "%ref\\\\\\\",[\\\\\\\"field-id%"]
+                                                     [:like :object "%ref\\\\\\\",[\\\\\\\"field-literal%"]
+                                                     [:like :object "%ref\\\\\\\",[\\\\\\\"fk->%"]]]}))))
+
+(define-reversible-migration RevisionAddJoinAliasToColumnSettingsFieldRefs
+  ;; This migration is essentially the same as `AddJoinAliasToColumnSettingsFieldRefs`, but for card revisions.
+  ;; We can't use the same migration because cards in the revision table don't always have `result_metadata`.
+  ;; So instead, we use the join aliases from card's `dataset_query` to create field refs in visualization_settings.
+  ;; There will inevitably be extra entries in visualization_settings.column_settings that don't match field refs in result_metadata, but that's ok.
+  (let [add-join-aliases
+        (fn [card]
+          (let [join-aliases (->> (get-in card ["dataset_query" "query" "joins"])
+                                  (map #(get % "alias"))
+                                  set)]
+            (if (seq join-aliases)
+              (update (get card "visualization_settings") "column_settings"
+                      (fn [column_settings]
+                        (let [copies-with-join-alias (into {}
+                                                           (mapcat (fn [[k v]]
+                                                                     (match (vec (json/parse-string k))
+                                                                       ["ref" ["field" id opts]]
+                                                                       (for [alias join-aliases]
+                                                                         [(json/generate-string ["ref" ["field" id (assoc opts "join-alias" alias)]]) v])
+                                                                       _ '()))
+                                                                   column_settings))]
+                          ;; existing column settings should take precedence over the copies in case there is a conflict
```

Comment:
Small detail: the case of a conflict should only be possible if the revision happened since the `AddJoinAliasToVisualizationSettingsFieldRefs` migration (v47.00-028), but we handle it anyway just in case